src/main/scala/io/citrine/lolo/trees/multitask/MultiTaskTrainingNode.scala

```diff
@@ -2,171 +2,207 @@ package io.citrine.lolo.trees.multitask
+
+      if (reducedData.isEmpty) {
+        Inaccessible()
+      } else if (split.isInstanceOf[NoSplit] || reducedData.length <= minInstances) {
+        val trainingLeaf: TrainingNode[AnyVal] = if (exampleLabel.isInstanceOf[Double]) {
+          RegressionTrainingLeaf
+            .build(
+              reducedData.asInstanceOf[Seq[(Vector[AnyVal], Double, Double)]],
+              GuessTheMeanLearner(),
+              maxDepth - remainingDepth,
+              rng
+            )
         } else {
-          new TrainingLeaf[Char](
-            reducedData.asInstanceOf[Seq[(Vector[AnyVal], Char, Double)]],
-            GuessTheMeanLearner(),
-            1,
-            rng = Random()
-          ).getNode()
+          ClassificationTrainingLeaf
+            .build(
+              reducedData.asInstanceOf[Seq[(Vector[AnyVal], Char, Double)]],
+              GuessTheMeanLearner(),
+              maxDepth - remainingDepth,
+              rng
+            )
         }
+        Stop(trainingLeaf)
+      } else if (left.nonEmpty && right.nonEmpty) {
+        DoSplit()
+      } else if (left.nonEmpty) {
+        GoLeft()
+      } else {
+        GoRight()
+      }
+    }
+
+    split match {
+      case split: NoSplit =>
+        MultiTaskTrainingNode(
+          trainingData = trainingData,
+          labelWiseInstructions = labelWiseInstructions,
+          split = split,
+          leftNodeOption = None,
+          rightNodeOption = None,
+          deltaImpurity = deltaImpurity
+        )
+      case split: Split =>
+        val (leftTrain, rightTrain) = trainingData.partition(r => split.turnLeft(r._1))
+        val leftNode = MultiTaskTrainingNode.build(
+          trainingData = leftTrain,
+          numFeatures = numFeatures,
+          remainingDepth = remainingDepth - 1,
+          maxDepth = maxDepth,
+          minInstances = minInstances,
+          splitter = splitter,
+          rng = rng
+        )
+        val rightNode = MultiTaskTrainingNode.build(
+          trainingData = rightTrain,
+          numFeatures = numFeatures,
+          remainingDepth = remainingDepth - 1,
+          maxDepth = maxDepth,
+          minInstances = minInstances,
+          splitter = splitter,
+          rng = rng
+        )
+        MultiTaskTrainingNode(
+          trainingData = trainingData,
+          labelWiseInstructions = labelWiseInstructions,
+          split = split,
+          leftNodeOption = Some(leftNode),
+          rightNodeOption = Some(rightNode),
+          deltaImpurity = deltaImpurity
+        )
     }
   }
 }
+
+/**
+  * An enumeration specifying what behavior to perform for a specific label.
+  * The logic of MultiTaskTrainingNode.build() enforces that instructions will only be given in valid situations.
+  * For example, DoSplit() will only occur if there is a valid split and this label has training data that
+  * go down both paths.
+  */
+sealed trait MultiTaskLabelInstruction
```

Comment:
I have personally developed the taste of organizing ADTs/enums within the companion object of the class they're associated with. So this could be added within the above object for `MultiTaskTrainingNode`. But it's a personal preference thing so not a blocker.

Reply:
I know you do, but I like them as separate traits/case classes